### ISSUE TITLE
Use dynamic jetton lock period

### ISSUE THREAD

**Issue by ya7on** [supadupadao/skipper#20]
At the moment `Jetton Lock` contract always prolong lock expiration date by constant value of 30 days on proposal voting.

This creates conditions in which every Proposal voting limited by 30 days. If we want to create proposal with voting duration of (for example) 3 month we have to change lock duration to that value for every(!) proposal, we have to change global constant value.
It obliges DAO participant to lock their tokens to 3 month even if they didn't want to participate so long votings.

I suggest to use dynamic lock period and dynamic proposal voting period. It will allow to create proposals with any duration. But it imposes limit to participants who want to vote in this proposal. Proposal should check lock duration before accept participant vote.

**Comment by iamIcarus** [comment]
atm with this lock structure , the DAO doesn't support participation in multiple proposals simultaneously, is that correct ?

**Comment by ya7on** [comment]
It support multiple proposals simultaneously, but it doesn't support dynamic lock time. Every proposal is hardcoded with same voting time. Lock time is changing globally not dynamically for every proposal

**Comment by iamIcarus** [comment]
apologies, just to clarify , if i hold max 100 jettons and lock all of them all to vote for a proposal, during that lock time i wont be able to participate in another proposal as a voter, or is this not the case?

**Comment by ya7on** [comment]
> this not the case

No it's not

With or without changes described in this issue there is opportunity to vote in multiple proposals.

At first you lock your jettons in Lock contract and after it become proxy for voting messages. So you send vote message to Lock, it proxies message to Proposal contract so Proposal could approve that you really have staked jettons. And after "Vote" message lock prolong staking time to 30d after current moment.

And in issue i tried to describe a way to prolong Lock staking time not to hardcoded period, but to period needed for specific Proposal.

**Comment by ya7on** [comment]
I tried to draw messages pipeline for voting [here](https://github.com/supadupadao/skipper/blob/master/README.md#vote-for-existing-proposal). It's not detailed unfortunately.

**Comment by iamIcarus** [comment]
got it thank you!

what do you think about adding an optional cooldown period to the unstake op (when the dynamic lock is implemented)? This is to prevent abuse or sales immediately after a proposal vote is over

**Comment by ya7on** [comment]
This feature is for more flexible customization of proposals. For example if DAO participant want to create half year-long voting, but default voting time is 30 days

**Comment by iamIcarus** [comment]
If we use a dynamic lock period, there will be the need to keep increasing it in order to participate in multiple proposals that could possibly be created in different times and for different voting timeframes 

how about:

- lock them without time period (indefinitely) until the user choose to unlock
- unlock can only happen if the user is not a voter in an active proposal.
- unlock can have an optional cooldown period before receiving the jettons

**Comment by ya7on** [comment]
> If we use a dynamic lock period, there will be the need to keep increasing it in order to participate in multiple proposals that could possibly be created in different times and for different voting timeframes

Yes, you are right. Lock period should be increased on every vote.

> unlock can have an optional cooldown period before receiving the jettons

Is this cooldown is some kind of global parameters? Where is it set? And what purpose of setting this parameter?

**Comment by iamIcarus** [comment]
cooldown period is to prevent abusive behaviors like:

> Lock minutes before proposal expire, vote , unlock and sell jettons as soon as the proposal is executed

This is ofc an addition to the core logic and can come at a later development stage. I can see this being linked to the unlock logic where it should prevent the user to unlock if user is participant to any active proposal. 

It could be an extra attribute in the proposal contract that sets a cooldown period before giving the ok to unlock.

Hope it makes sense

**Comment by ya7on** [comment]
> Lock minutes before proposal expire, vote , unlock and sell jettons as soon as the proposal is executed

Never really thought about that case. Thank you highlighting it. What kind of abuse can potentially be done without additional lock? What immediately sale can lead to?

**Comment by iamIcarus** [comment]
> What immediately sale can lead to

Depending on the amount of jettons the user holds (and if the jettons are traded in a dex) can fluctuate the price significantly. 


A benefit we get from moving from a simple time lock process for voting, to a process that involves each proposal contract (the user participated) to give the ok for unlocking, is that the unlock logic can be easily enhanced later.

The cost is that its a more complex implementation

What's your preferred way of handling this?

**Comment by ya7on** [comment]
I got it. Thank you for explanation.

> The cost is that its a more complex implementation
> What's your preferred way of handling this?

Honestly, don't know. I'll open issue for additional cooldown, because i think it's not related with this issue topic. But seems like it's important feature.

**Comment by iamIcarus** [comment]
@ya7on i'll try to pick this soon and implement it by increasing the lock period (if needed) each time the user tries to participate in a new proposal. Lets keep it simple for now 😀

**Comment by ya7on** [comment]
Actually lock period increasing to static period is already implemented. Here is line of code where it implemented:
https://github.com/supadupadao/skipper/blob/315db48aa79219309d963a4e6c1a7dcca288875b/contracts/lock.tact#L40
Or did I misunderstand and you meant something else?

**Comment by iamIcarus** [comment]
im looking to improve that by using the proposal contract expiry date to calculate if lock extension is needed instead of always using the constant value of `LOCK_INTERVAL`

This way lock period will extend to cover the period of the contract the user participates that will expire last 

I still didn’t get a chance to look into the skipper in detail so apologies if this is already implemented

**Comment by ya7on** [comment]
> if this is already implemented

Sorry, it doesn't implemented. I thought about another thing.
Good point! Your improvement is really needed

**Comment by iamIcarus** [comment]
do you think this implementation will cover this issue or do you think additional or different approach is needed?

**Comment by ya7on** [comment]
Looks like it will cover issue on Jetton Lock side, but i think there also need to implement lock period setting for Proposal

**Comment by ya7on** [comment]
Resolved in #29
